### ISSUE TITLE
Function return type does not match operand type of return inst!

### ISSUE THREAD

**Issue by riffraff169** [crystal-lang/crystal#10544]
## Bug Report
This code compiles perfectly fine under 0.35.1, but fails with any version later (such as 0.36.0 up to 1.0.0).

Minimal code:
```
alias Lexeme = Hash(Symbol, String|Int32|Symbol)

class L1
  def p1
    lexemes = [] of Lexeme
    if (1)
      lexeme = {:type => :pragma, :value => 0, :line => 1}
    elsif (2)
      lexeme = {:type => :pragma, :value => :token, :line => 2}
    else
      lexeme = {:type => :pragma, :value => "string", :line => 3}
    end
    lexemes << lexeme
  end
end

l1 = L1.new
lex1 = [] of Lexeme
lex2 = l1.p1
```

Here is stack trace:

```
Module validation failed: Function return type does not match operand type of return inst!
  ret i32* %1, !dbg !36
 %"Hash(Symbol, Int32 | String | Symbol)"* (Exception)
  from /crystal/src/int.cr:538:9 in 'finish'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:69:7 in 'codegen'
  from /crystal/src/compiler/crystal/compiler.cr:172:16 in 'compile'
  from /crystal/src/compiler/crystal/command.cr:296:3 in 'run'
  from /crystal/src/compiler/crystal.cr:11:1 in '__crystal_main'
  from /crystal/src/crystal/main.cr:110:5 in 'main'
  from src/env/__libc_start_main.c:94:2 in 'libc_start_main_stage2'
Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/crystal-lang/crystal/issues
```

**Comment by asterite** [comment]
As a workaround you can do this:

```crystal
alias Lexeme = Hash(Symbol, String | Int32 | Symbol)

class L1
  def p1
    lexemes = [] of Lexeme
    if (1)
      lexeme = Lexeme{:type => :pragma, :value => 0, :line => 1}
    elsif (2)
      lexeme = Lexeme{:type => :pragma, :value => :token, :line => 2}
    else
      lexeme = Lexeme{:type => :pragma, :value => "string", :line => 3}
    end
    lexemes << lexeme
  end
end

l1 = L1.new
lex1 = [] of Lexeme
lex2 = l1.p1
```

The original should shouldn't compile.

**Comment by straight-shoota** [comment]
Reduced:
```cr
lexemes = [] of Array(String|Int32)
lexemes << (true ? [0] : [0, "string"])
```

**Comment by straight-shoota** [comment]
This regression is already present in 0.36.1. It was introduced in #10081.

**Comment by asterite** [comment]
I don't know if it's a regression or not, but it should never compile.

**Comment by straight-shoota** [comment]
The regression is that the code previously resulted in a proper error message. Now we get an ICE instead.

**Comment by straight-shoota** [comment]
Ignore me. It wasn't an error before and that's obviously wrong.

**Comment by riffraff169** [comment]
Yeah, sorry, I hadn't tested the build since 0.35.1... been kind of idle.

**Comment by HertzDevil** [comment]
Reduced:

```crystal
class Foo(T)
end

class Bar1
end

class Bar2
end

buffer = Pointer(Foo(Bar1 | Bar2)).null
buffer[0] = buffer[0]
buffer.value = Foo(Bar1).new || Foo(Bar1 | Bar2).new
```

```
Module validation failed: Function return type does not match operand type of return inst!
  ret i32* %1, !dbg !11
 %"Foo(Bar1 | Bar2)"* (Exception)
  from /crystal/src/int.cr:538:9 in 'finish'
  from /crystal/src/compiler/crystal/codegen/codegen.cr:69:7 in 'codegen'
  from /crystal/src/compiler/crystal/compiler.cr:172:16 in 'compile'
  from /crystal/src/compiler/crystal/command.cr:210:5 in 'run_command'
  from /crystal/src/compiler/crystal/command.cr:117:7 in 'run'
  from /crystal/src/compiler/crystal.cr:11:1 in '__crystal_main'
  from /crystal/src/crystal/main.cr:110:5 in 'main'
  from src/env/__libc_start_main.c:94:2 in 'libc_start_main_stage2'
```

If we change the second `buffer[0]` to `buffer.value` a different error shows up:

```
Module validation failed: Call parameter type does not match function signature!
  %59 = load i32*, i32** %58, !dbg !76
 %"Foo(Bar1 | Bar2)"*  %60 = call %"Foo(Bar1 | Bar2)"* @"*Pointer(Foo(Bar1 | Bar2) | Foo(Bar1))@Pointer(T)#[]=<Int32, Foo(Bar1 | Bar2)>:Foo(Bar1 | Bar2)"(i32** %57, i32 0, i32* %59), !dbg !77
 (Exception)
```

The full fix is simply #10214, but a more targeted fix for `Pointer#value=` is also possible.
